File: sources/r/03_func_sum/main.R
# task 03 func_sum — expected output: 100000000
# build: Rscript main.R    run: Rscript main.R
#
# R has no no-inline attribute, and it has no inliner either: a call to a
# closure is a real call through the closure every time, so a plain function
# is already a hundred million genuine calls.

add_one <- function(n) n + 1

func_sum <- function() {
  value <- 0
  i <- 0
  while (i < 100000000) {
    value <- add_one(value)
    i <- i + 1
  }
  cat(sprintf("%.0f\n", value))
}

func_sum()

File: sources/python/03_func_sum/main.py
# task 03 func_sum — expected output: 100000000
# build: python main.py | pypy main.py | graalpy main.py | nuitka --standalone main.py    run: python main.py
# note: CPython never inlines this call, so the 100000000 calls really happen; PyPy and the nuitka build may inline it away, since plain Python has no no-inline attribute.

def add_one(n):
    return n + 1


value = 0
for _ in range(100000000):
    value = add_one(value)

print(value)
